### PR TITLE
Hotplug the cpus in random order by selecting random core.

### PR DESCRIPTION
**Pull request by shriyak**
The script will select random core from the list, picks the random cpus and does
offline and online in random order for that core.Make sure that all cores
are online.Script will set smt to maximum value,later it
works in random order to offline and online the cpu for 100 iterations.

Signed-off-by: Shriya <shriyak@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by PraveenPenguin** [comment]
@abdhaleegit can you please help us to review this

**Comment by abdhaleegit** [comment]
@shriyakul a similar scenario is already covered in cpustress.py

File: cpu/cpu_hotplug.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        Offline the particular cpu
+        """
+        cmd = "echo 0 > %s/cpu%s/online" % (self.path, cpu_num)
+        process.system(cmd, shell=True)
+        time.sleep(1)
+        cmd = "cat %s/cpu%s/online" % (self.path, cpu_num)
+        val = process.system_output(cmd, shell=True)
+        if val != '0':
+            self.nfail += 1
+            self.log.info("Failed to offline the cpu %s" % cpu_num)
```
Comment (abdhaleegit) [inline cpu/cpu_hotplug.py]: There is an existing Avocado api which take cares of this utils/cpu.py : 
from avocado.utils import cpu
cpu.offline(cpu)
cpu.online(cpu)
and more

File: cpu/cpu_hotplug.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    def random_gen_cores(self):
+        """
+        Generate random core list
+        """
+        nums = [x for x in range(0, self.T_CORES)]
+        random.shuffle(nums)
+        self.log.info(" Core list is %s" % nums)
+        return nums
+
```
Comment (abdhaleegit) [inline cpu/cpu_hotplug.py]: To generate a random number you can use 
from random import randint
core = randint(0, cores)

File: cpu/cpu_hotplug.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    def random_gen_cpu(self, core):
+        """
+        Generate random cpu number for the given core
+        """
+        nums = [x for x in range(self.max_smt * core,
+                ((self.max_smt * core) + self.max_smt))]
+        random.shuffle(nums)
+        return nums
```
Comment (abdhaleegit) [inline cpu/cpu_hotplug.py]: cpu =  randint(cpus)

File: cpu/cpu_hotplug.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def offline_cpu(self, cpu_num):
+        """
+        Offline the particular cpu
+        """
+        cmd = "echo 0 > %s/cpu%s/online" % (self.path, cpu_num)
+        process.system(cmd, shell=True)
+        time.sleep(1)
```
Comment (abdhaleegit) [inline cpu/cpu_hotplug.py]: remove time.sleep

File: cpu/cpu_hotplug.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        self.THREADS = process.system_output("lscpu | grep 'Thread(s) per core"
+                                             ":'| awk '{print $4}'",
+                                             shell=True)
+        self.T_CORES = int(self.CORES) * int(self.SOCKETS)
+        self.log.info(" Cores = %s and threads = %s "
+                      % (self.T_CORES, self.THREADS))
+        process.system_output("echo 8 > /proc/sys/kernel/printk", shell=True)
+        self.max_smt = 4
+        if cpu.get_cpu_arch().lower() == 'power8':
+            self.max_smt = 8
```
[the remark this answers is not in the document]
Comment (shriyak) [reply cpu/cpu_hotplug.py]: it is default set to P9 value, smt=4.

File: cpu/cpu_hotplug.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            self.cancel("Processor is not ppc64")
+        self.nfail = 0
+        self.CORES = process.system_output("lscpu | grep 'Core(s) per socket:'"
+                                           "| awk '{print $4}'", shell=True)
+        self.SOCKETS = process.system_output("lscpu | grep 'Socket(s):'"
+                                             "| awk '{print $2}'", shell=True)
+        self.THREADS = process.system_output("lscpu | grep 'Thread(s) per core"
+                                             ":'| awk '{print $4}'",
+                                             shell=True)
+        self.T_CORES = int(self.CORES) * int(self.SOCKETS)
```
[the remark this answers is not in the document]
Comment (shriyak) [reply cpu/cpu_hotplug.py]: Sure, let me make this change

File: cpu/cpu_hotplug.py (hunk 1 above)
Comment (shriyak) [reply cpu/cpu_hotplug.py]: thanks , I will update it

**Comment by shriyak** [comment]
@abdhaleegit , @PraveenPenguin  : This script is  for testing EM , which will enhanced later at quad level. Let me make the changes and update.

File: cpu/cpu_hotplug.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shriyak) [reply cpu/cpu_hotplug.py]: Here , I need total cores not cpus, so I do sockets * cores = total-cores. so 4 cores considering the above example

File: cpu/cpu_hotplug.py (hunk 2 above)
Comment (shriyak) [reply cpu/cpu_hotplug.py]: I need list here, so I use the above logic which gives shuffled core list.
Ex : 
[stdout] [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
 Core list is [22, 18, 8, 11, 0, 12, 2, 1, 23, 7, 10, 17, 19, 6, 20, 5, 4, 13, 14, 15, 16, 3, 9, 21]

File: cpu/cpu_hotplug.py (hunk 3 above)
Comment (shriyak) [reply cpu/cpu_hotplug.py]: Here also same, I need cpu list which is shuffled for each core.

Ex : [88, 90, 91, 89] for the core : 22

**Comment by shriyak** [comment]
closing this PR , as I am raising new one.

**Pull request closed without merging**